File: src/jira_dc_mcp/tools/filters.py
"""Filter, dashboard, and webhook introspection tools."""

from __future__ import annotations

import json

from ..client import JiraClient


async def list_filters(client: JiraClient) -> str:
    """List favourite/shared JQL filters."""
    filters = await client.list_filters()
    result = [
        {
            "id": f.get("id"),
            "name": f.get("name"),
            "jql": f.get("jql"),
            "owner": (f.get("owner") or {}).get("displayName"),
            "favourite": f.get("favourite"),
            "favouritedCount": f.get("favouritedCount"),
            "sharePermissions": [
                {
                    "type": sp.get("type"),
                    "project": (sp.get("project") or {}).get("key") if sp.get("project") else None,
                    "role": (sp.get("role") or {}).get("name") if sp.get("role") else None,
                    "group": (sp.get("group") or {}).get("name") if sp.get("group") else None,
                }
                for sp in f.get("sharePermissions", [])
            ],
        }
        for f in filters
    ]
    return json.dumps(result, indent=2)


async def list_dashboards(client: JiraClient) -> str:
    """List all dashboards."""
    dashboards = await client.list_dashboards()
    result = [
        {
            "id": d.get("id"),
            "name": d.get("name"),
            "owner": (d.get("owner") or {}).get("displayName") if d.get("owner") else None,
            "popularity": d.get("popularity"),
            "view": d.get("view"),
        }
        for d in dashboards
    ]
    return json.dumps(result, indent=2)


async def list_webhooks(client: JiraClient) -> str:
    """List all registered webhooks."""
    webhooks = await client.list_webhooks()
    result = [
        {
            "id": w.get("self", "").rstrip("/").split("/")[-1] if w.get("self") else w.get("id"),
            "name": w.get("name"),
            "url": w.get("url"),
            "events": w.get("events", []),
            "enabled": w.get("enabled"),
            "filters": w.get("filters") or {},
            "excludeBody": w.get("excludeBody"),
        }
        for w in webhooks
    ]
    return json.dumps(result, indent=2)

File: src/tools/filters.ts
/** Filter and dashboard introspection tools. */

import { dumps } from "../json.js";
import type { ToolDef } from "./types.js";
import { filterByName, nameFilterShape, pageShape, paginate } from "./util.js";

const DASHBOARD_PAGE = 200;

export const filterTools: ToolDef[] = [
  {
    name: "list_filters",
    description:
      "List favourite/shared JQL filters visible to the authenticated user. " +
      "Shows filter name, JQL query, owner, and share permissions.",
    inputShape: {},
    async handler({ client }) {
      const filters = await client.listFilters();
      return dumps(
        filters.map((f: any) => ({
          id: f.id,
          name: f.name,
          jql: f.jql,
          owner: (f.owner ?? {}).displayName,
          favourite: f.favourite,
          favouritedCount: f.favouritedCount,
          sharePermissions: (f.sharePermissions ?? []).map((sp: any) => ({
            type: sp.type,
            project: sp.project ? (sp.project.key ?? null) : null,
            role: sp.role ? (sp.role.name ?? null) : null,
            group: sp.group ? (sp.group.name ?? null) : null,
          })),
        })),
      );
    },
  },

  {
    name: "list_dashboards",
    description:
      "List dashboards visible to the authenticated user: id, name, and view URL. " +
      "The dashboard REST API does not expose owner or popularity. " +
      "Paginated (offset/limit) and filterable by name_contains.",
    inputShape: { ...nameFilterShape, ...pageShape(DASHBOARD_PAGE) },
    async handler({ client }, args) {
      const dashboards = filterByName(await client.listDashboards(), args.name_contains);
      return dumps(
        paginate(
          dashboards.map((d: any) => ({ id: d.id, name: d.name, view: d.view })),
          args,
          DASHBOARD_PAGE,
        ),
      );
    },
  },
];
